### PR TITLE
Configuration canary for the second or further plugins.

### PR DESCRIPTION
**Pull request by ingwonsong**
Fixes: https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/175

Applying Configuration Canary for second or further plugin.

In https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/175, there are some discussion about 1) needs of optimization for multiple plugins, and 2) the initialization in the main thread.

But, before making such optimization, this PR proposes to solve the issue itself because it seems to be critical in terms of stability.

### CODE REVIEW

**Comment by ingwonsong** [comment]
@mathetake @PiotrSikora Friendly ping.

**Comment by PiotrSikora** [comment]
Can you add tests?

Also, the original issue was a single plugin with multiple configurations (one valid, one invalid), but this is talking about multiple plugins, so it's unclear which case you are trying to address.

**Comment by ingwonsong** [comment]
> Also, the recent issue was a single plugin with multiple configurations (one valid, one invalid), but this is talking about multiple plugins, so it's unclear which case you are trying to address.

I thought that the multiple filters with the same Wasm was problematic: https://github.com/envoyproxy/envoy/pull/20843#issuecomment-1101240349

In my understanding, `Plugin` is created per `filter config` (per configuration), while `WasmVM` is created per `vm_key`. Isn't it right?

**Comment by PiotrSikora** [comment]
> I thought that the multiple filters with the same WasmVM was problematic: [envoyproxy/envoy#20843 (comment)](https://github.com/envoyproxy/envoy/pull/20843#issuecomment-1101240349)
> 
> In my understanding, `Plugin` is created per `filter config` (per configuration), while `Wasm base handle` (before cloning per thread) is created per `vm_key`. Isn't it right?

That's the issue with different instances of the same Wasm plugin (i.e. when same was plugin is started with different plugin configuration).

However, #175 (which this PR supposedly resolves) is about multiple Wasm plugins linked in the same Wasm module, which I don't belive this fixes.

**Comment by ingwonsong** [comment]
Oh.. "multiple Wasm plugin" means a kind of multiple type? (like network, http, bootstrap)

According to https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/175#issue-931110544, #175 seems to be about that "canary" is not applied to the second plugin instance when `vm_key` is the same.

**Comment by PiotrSikora** [comment]
@ingwonsong you need to add rule to build `wasm_cc_binary` targets in `test_data` step on the CI (`.github/workflows/test.yml`).

@chaoqin-li1123 @mpwarres could you do the first review pass? Thank you.

**Review by PiotrSikora — COMMENTED** [review]
@ingwonsong could you update Proxy-Wasm C++ SDK to latest as part of this PR?

File: src/wasm.cc [hunk 1]
```diff
@@ -446,6 +446,32 @@ void WasmBase::finishShutdown() {
   }
 }
 
+bool canary(std::shared_ptr<WasmHandleBase> &wasm_handle, const std::shared_ptr<PluginBase> &plugin,
```
Comment (chaoqin-li1123) [inline src/wasm.cc:449]: Does it make sense to make canary(...) a member function of WasmHandle?

Comment (ingwonsong) [reply src/wasm.cc:449]: That makes sense. I have changed it.

File: src/wasm.cc [hunk 2]
```diff
@@ -447,6 +447,32 @@ void WasmBase::finishShutdown() {
   }
 }
 
+bool WasmHandleBase::canary(const std::shared_ptr<PluginBase> &plugin,
+            const WasmHandleCloneFactory &clone_factory) {
+  auto configuration_canary_handle = clone_factory(shared_from_this());
+  if (!configuration_canary_handle) {
+    this->wasm()->fail(FailState::UnableToCloneVm, "Failed to clone Base Wasm");
```
Comment (mpwarres) [inline src/wasm.cc:454]: AIUI this is calling fail() on a WasmBase instance that may already be in use by an earlier caller of createWasm()--does that raise any issues? @PiotrSikora 

Maybe in practice this is very unlikely, since that earlier caller of createWasm() would have gone through this codepath as well, so if clone_factory were broken, it would have failed then as well.

File: src/wasm.cc [hunk 3]
```diff
@@ -465,44 +491,29 @@ std::shared_ptr<WasmHandleBase> createWasm(const std::string &vm_key, const std:
         base_wasms->erase(it);
```
Comment (mpwarres) [inline src/wasm.cc:491]: This isn't a new thing, but I wonder if there is anything that would clear a failed WasmBase out of the base_wasms map.  With the local_wasms map, a fail callback is set up to clear out a map entry in case of failure ([code](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/2cf68d6c00b4141ecf2afb5e5d51ebd21a27ed43/src/wasm.cc#L552)), but I don't see anything similar for base_wasms.

File: test/utility.h [hunk 4]
```diff
@@ -118,10 +128,37 @@ class TestContext : public ContextBase {
...
+           std::string_view vm_configuration, std::string_view vm_key,
+           std::unordered_map<std::string, std::string> envs,
+           AllowedCapabilitiesMap allowed_capabilities)
+      : WasmBase(std::move(wasm_vm), vm_id, vm_configuration, vm_key, std::move(envs),
+                 std::move(allowed_capabilities)) {}
+
+  TestWasm(std::unique_ptr<WasmVm> wasm_vm, std::unordered_map<std::string, std::string> envs)
+      : TestWasm(std::move(wasm_vm), "", "", "", std::move(envs), {}) {}
+
+  TestWasm(std::unique_ptr<WasmVm> wasm_vm) : TestWasm(std::move(wasm_vm), {}) {}
```
Comment (mpwarres) [inline test/utility.h:141]: Is the 1-arg unique_ptr<WasmVm> constructor used anywhere?

(optional) 4 constructors seems like a lot. I wonder if you could get away with just 2, i.e. the 6-arg unique_ptr<WasmVm> one and the 3-arg shared_ptr<WasmHandleBase> one.

**Comment by ingwonsong** [comment]
> @ingwonsong could you update Proxy-Wasm C++ SDK to latest as part of this PR?

Done.

**Comment by ingwonsong** [comment]
> @ingwonsong you need to add rule to build `wasm_cc_binary` targets in `test_data` step on the CI (`.github/workflows/test.yml`).

Done, too.

File: src/wasm.cc (hunk 2 above)
Comment (ingwonsong) [reply src/wasm.cc:454]: I am not 100% sure, but AFAIK, the earlier caller of the createWasm() should rely on the thread-specific Wasm VM. Therefore, I think that the earlier caller is not affected by calling "fail()" later. 
@PiotrSikora WDYT?

(BTW, as we know, this is the code piece which was already here.)

File: src/wasm.cc (hunk 3 above)
Comment (ingwonsong) [reply src/wasm.cc:491]: I agree. 

But, thinking about the evolution of codes, IMHO, this kind of defensive codes seems to be good.

File: test/utility.h (hunk 4 above)
Comment (ingwonsong) [reply test/utility.h]: 1-arg unique_ptr is used in the existing codes.
(e.g., https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/2cf68d6c00b4141ecf2afb5e5d51ebd21a27ed43/test/exports_test.cc#L77)

BTW, I reduced them from 5 to 3. 
2-arg is also possible, but 3-arg seems to be slightly cleaner when using TestWasm class in test codes.

File: test/test_data/configure_check.cc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#include <unordered_map>
+
+#include "proxy_wasm_intrinsics.h"
+
+class TestRootContext1 : public RootContext {
+public:
+  explicit TestRootContext1(uint32_t id, std::string_view root_id) : RootContext(id, root_id) {}
+  bool onConfigure(size_t s) override {
+    LOG_TRACE("onConfigure: root_id_1");
+    return true;
```
Comment (mpwarres) [inline test/test_data/configure_check.cc:26]: It might be good for the new TestVm.AlwaysApplyCanary test to exercise the case of canary failure. One way to do that could be for the onConfigure() method here to return false depending on the value of the configuration. Then the test case could feed configuration values that provoke canary failure (or not).

File: test/test_data/configure_check.cc [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "proxy_wasm_intrinsics.h"
+
+class TestRootContext1 : public RootContext {
```
Comment (mpwarres) [inline test/test_data/configure_check.cc:21]: Is there value in having separate pairs of Test*Context1 vs. Test*Context2 classes? Alternatively this could just define a single pair of classes, TestRootContext and TestContext, and have onConfigure log the constructor-provided root_id value. There would still be two RegisterContextFactory instantiations with "root_id_1" and "root_id_2", but they would both use CONTEXT_FACTORY(TestContext) and ROOT_FACTORY(TestRootContext).

Or am I misunderstanding and this would sacrifice some aspect of the test?

File: test/wasm_test.cc [hunk 7]
```diff
@@ -113,4 +113,122 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
   ASSERT_NE(thread_local_plugin3->wasm(), thread_local_plugin2->wasm());
 }
 
-} // namespace proxy_wasm
+// Tests the canary is always applied when making a call `createWasm`
+TEST_P(TestVm, AlwaysApplyCanary) {
```
Comment (mpwarres) [inline test/wasm_test.cc:117]: Nice test!

**Review by mpwarres — COMMENTED** [review]
LGTM mod remaining questions.

File: test/test_data/configure_check.cc (hunk 6 above)
Comment (ingwonsong) [reply test/test_data/configure_check.cc:21]: You are right. 

I thought that using the same context with the different root context is unusual, so, for the demonstration purpose,  I used the separate contexts: `TestContext1` and `TestContext2`. 

BTW, Done as you commented.

File: test/test_data/configure_check.cc (hunk 5 above)
Comment (ingwonsong) [reply test/test_data/configure_check.cc:29]: Good point. Done!

File: test/test_data/configure_check.cc [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+public:
+  explicit TestRootContext1(uint32_t id, std::string_view root_id) : RootContext(id, root_id) {}
+  bool onConfigure(size_t s) override {
+    LOG_TRACE("onConfigure: root_id_1");
+    if (s == 0) {
```
Comment (mpwarres) [inline test/test_data/configure_check.cc:26]: Shorten to:
`return s != 0;`

File: test/test_data/configure_check.cc [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+public:
+  explicit TestContext(uint32_t id, RootContext *root) : Context(id, root) {}
+};
+
+class TestRootContext2 : public RootContext {
+public:
+  explicit TestRootContext2(uint32_t id, std::string_view root_id) : RootContext(id, root_id) {}
+  bool onConfigure(size_t s) override {
+    LOG_TRACE("onConfigure: root_id_2");
+    if (s == 0) {
```
Comment (mpwarres) [inline test/test_data/configure_check.cc:43]: Shorten to:
`return s != 0;`

File: test/wasm_test.cc [hunk 10]
```diff
@@ -113,4 +113,131 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
...
+            // Create a base Wasm by createWasm.
+            auto wasm_handle_comp =
+                createWasm(vm_key, source, plugin_comp, wasm_handle_factory_comp,
+                           wasm_handle_clone_factory_for_canary, false);
+
+            if (plugin_config.empty()) {
+              // configure_check.wasm should raise the error at `onConfigure` in canary when the
+              // `plugin_config` is empty string.
+              EXPECT_EQ(wasm_handle_comp, nullptr);
+              return;
```
Comment (mpwarres) [inline test/wasm_test.cc:208]: Should this be `continue` rather than `return`? ISTM `return` would cut short the entire test case.

**Review by mpwarres — COMMENTED** [review]
Remaining comments mostly optional, with the exception of the `return` vs. `continue` one.

File: test/wasm_test.cc (hunk 10 above)
Comment (ingwonsong) [reply test/wasm_test.cc:208]: Oops.. Thanks for checking this!

**Review by mpwarres — APPROVED** [review]
LG, thanks!

**Comment by ingwonsong** [comment]
@PiotrSikora Wam reminder. :)

File: test/test_data/configure_check.cc [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    LOG_TRACE("onConfigure: root_id_2");
+    return s != 0;
+  }
+};
+
+static RegisterContextFactory register_TestContext1(CONTEXT_FACTORY(TestContext),
+                                                    ROOT_FACTORY(TestRootContext1), "root_id_1");
+
+static RegisterContextFactory register_TestContext2(CONTEXT_FACTORY(TestContext),
+                                                    ROOT_FACTORY(TestRootContext2), "root_id_2");
```
Comment (PiotrSikora) [inline test/test_data/configure_check.cc:48]: Nit: missing newline at the end of the file.

File: test/wasm_test.cc [hunk 12]
```diff
@@ -113,4 +113,131 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
...
+              wasm_handle_comp->kill();
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+} // namespace proxy_wasm
```
Comment (PiotrSikora) [inline test/wasm_test.cc:243]: Nit: missing newline at the end of the file.

File: .github/workflows/test.yml [hunk 13]
```diff
@@ -56,6 +56,17 @@ jobs:
...
         -c opt
         $(bazel query 'kind(was.*_rust_binary, //test/test_data/...)')
+        $(bazel query 'kind(_optimized_wasm_cc_binary, //test/test_data/...)')
+
+    - name: Clean up test data
+      shell: bash
+      run: |
+        # Remove temporal directories
+        for i in $(cd bazel-bin/test/test_data && ls -1); do \
+          if [ -d "bazel-bin/test/test_data/$i" ]; then \
```
Comment (PiotrSikora) [inline .github/workflows/test.yml:66]: Consider replacing this with `find bazel-bin/test/test_data/ -mindepth 1 -maxdepth 1 -type d`

File: .github/workflows/test.yml [hunk 14]
```diff
@@ -56,6 +56,17 @@ jobs:
         --config=clang
         -c opt
         $(bazel query 'kind(was.*_rust_binary, //test/test_data/...)')
+        $(bazel query 'kind(_optimized_wasm_cc_binary, //test/test_data/...)')
+
+    - name: Clean up test data
```
Comment (PiotrSikora) [inline .github/workflows/test.yml:61]: Why do we need to do it? Are the runfiles corrupted due to changing between compilation modes? If so, it might be worth adding the comment explaining why.

File: test/utility.h [hunk 15]
```diff
@@ -114,14 +132,26 @@ class TestContext : public ContextBase {
...
 };
 
 class TestWasm : public WasmBase {
 public:
   TestWasm(std::unique_ptr<WasmVm> wasm_vm, std::unordered_map<std::string, std::string> envs = {})
       : WasmBase(std::move(wasm_vm), "", "", "", std::move(envs), {}) {}
 
+  TestWasm(std::unique_ptr<WasmVm> wasm_vm, std::string_view vm_id,
+           std::string_view vm_configuration, std::string_view vm_key)
+      : WasmBase(std::move(wasm_vm), vm_id, vm_configuration, vm_key, {}, {}) {}
```
Comment (PiotrSikora) [inline test/utility.h:145]: Nit: consider merging it with the constructor above.

File: .github/workflows/test.yml (hunk 14 above)
Comment (ingwonsong) [reply .github/workflows/test.yml:61]: When using `proxy_wasm_cc_binary`, the final artifact seems to have `_optimized_wasm_cc_binary` kind.

```
$ bazel query //test/test_data/... --output=label_kind 2>&1 | grep configure_check
_optimized_wasm_cc_binary rule //test/test_data:configure_check.wasm
cc_binary rule //test/test_data:proxy_wasm_configure_check
wasm_cc_binary rule //test/test_data:wasm_configure_check.wasm
```

Will add the comment here.

File: .github/workflows/test.yml (hunk 13 above)
Comment (ingwonsong) [reply .github/workflows/test.yml:66]: Thanks! Done!

File: .github/workflows/test.yml (hunk 14 above)
Comment (PiotrSikora) [reply .github/workflows/test.yml:61]: Sorry, I didn't mean the `bazel query`, I asked about why do we need to clean up test data.

The comment about `_optimized_wasm_cc_binary` is pretty useless, feel free to remove it (and if not, then it should be in the correct step, not in the clean up).

Comment (ingwonsong) [reply .github/workflows/test.yml:64]: Done.

File: test/wasm_test.cc [hunk 16]
```diff
@@ -113,4 +113,131 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
   ASSERT_NE(thread_local_plugin3->wasm(), thread_local_plugin2->wasm());
 }
 
-} // namespace proxy_wasm
+// Tests the canary is always applied when making a call `createWasm`
+TEST_P(TestVm, AlwaysApplyCanary) {
+  // Skip this test for s309x arch because it tooks too much time in CI.
```
[the remark this answers is not in the document]
Comment (ingwonsong) [reply test/wasm_test.cc]: spawning new VM instance of wasmtime seems to be super slow. (+ s390x qemu overhead)
So, I changed the test to reuse VM.

**Comment by ingwonsong** [comment]
@PiotrSikora Can you take a look once again? :)

File: test/utility.h [hunk 17]
```diff
@@ -91,16 +91,34 @@ class TestIntegration : public WasmVmIntegration {
 class TestContext : public ContextBase {
 public:
   TestContext(WasmBase *wasm) : ContextBase(wasm) {}
+  TestContext(WasmBase *wasm, const std::shared_ptr<PluginBase> &plugin)
+      : ContextBase(wasm, plugin) {}
```
Comment (PiotrSikora) [inline test/utility.h:95]: It almost feels like we should have a separate `TestVmContext` and `TestRootContext` classes.

I'm going to allow this for now, beause all Context types are combined in the production code, and existing tests operate on VmContext, which doesn't match the production code or SDKs...

Something to keep in mind in the future, @mpwarres.

File: test/wasm_test.cc [hunk 18]
```diff
@@ -113,4 +114,132 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
...
+  const std::string root_ids[2] = {"root_id_1", "root_id_2"};
+  const std::string vm_ids[2] = {"vm_id_1", "vm_id_2"};
+  const std::string vm_configs[2] = {"vm_config_1", "vm_config_2"};
+  const std::string plugin_configs[3] = {"plugin_config_1", "plugin_config_2",
+                                         /* raising the error */ ""};
+  const std::string plugin_keys[2] = {"plugin_key_1", "plugin_key_2"};
+  const auto fail_open = false;
+
+  // Define common callbacks
+  WasmHandleCloneFactory wasm_handle_clone_factory =
```
Comment (PiotrSikora) [inline test/wasm_test.cc:131]: `wasm_handle_clone_factory` isn't used anywhere, and `wasm_handle_clone_factory_for_canary` is used twice. Is this is a leftover from copy&paste or should both versions be used?

File: test/wasm_test.cc [hunk 19]
```diff
@@ -113,4 +114,132 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
...
+            // plugin->key() is unique for root_id + plugin_config + plugin_key.
+            // plugin->key() is used as an identifier of local-specific plugins as well.
+            if (root_id == root_ids[0] && plugin_config == plugin_configs[0] &&
+                plugin_key == plugin_keys[0]) {
+              EXPECT_EQ(plugin_baseline->key(), plugin_comp->key());
+            } else {
+              EXPECT_NE(plugin_baseline->key(), plugin_comp->key());
+            }
+            // For each create Wasm, canary should be done.
+            EXPECT_EQ(canary_count, 1);
```
Comment (PiotrSikora) [inline test/wasm_test.cc:237]: I believe `canary_count` is increased with each clone, regardless of whether config canary was successful or not, so you might want to check it right after `createWasm`, or verify that `canary_count` matches total number of iterations. Checking it here doesn't really prove anything.

File: test/test_data/BUILD [hunk 20]
```diff
@@ -59,3 +60,8 @@ wasm_rust_binary(
     srcs = ["random.rs"],
     wasi = True,
 )
+
+proxy_wasm_cc_binary(
+    name = "configure_check.wasm",
+    srcs = ["configure_check.cc"],
```
Comment (PiotrSikora) [inline test/test_data/BUILD:66]: Nit: consider renamig to `canary_check`.

File: test/test_data/configure_check.cc [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "proxy_wasm_intrinsics.h"
+
+class TestRootContext1 : public RootContext {
```
Comment (PiotrSikora) [inline test/test_data/configure_check.cc:21]: There are already too many `Test*` classes, which adds a lot of noise when grepping... Could you rename this to something more meaningful? e.g. `CanaryCheckRootContext1`?

Same with other classess in this file.

File: test/wasm_test.cc [hunk 22]
```diff
@@ -15,6 +15,7 @@
 #include "include/proxy-wasm/wasm.h"
 
 #include "gtest/gtest.h"
+#include <unordered_set>
```
Comment (PiotrSikora) [inline test/wasm_test.cc:18]: Style: move `<unordered_set>` before `gtest/gtest.h` and add an empty line between them.

**Review by PiotrSikora — COMMENTED** [review]
LGTM, with a few small comments/nits.

File: test/wasm_test.cc (hunk 18 above)
Comment (ingwonsong) [reply test/wasm_test.cc:131]: Ah.. good point. I had to remove this one.

File: test/wasm_test.cc (hunk 19 above)
Comment (ingwonsong) [reply test/wasm_test.cc:237]: Moved to right after of the `creatWasm`.

File: test/wasm_test.cc (hunk 22 above)
Comment (ingwonsong) [reply test/wasm_test.cc:18]: Done

File: test/test_data/configure_check.cc (hunk 21 above)
Comment (ingwonsong) [reply test/test_data/configure_check.cc]: Done to change to use "CanaryCheck" prefix.

File: test/test_data/BUILD (hunk 20 above)
Comment (ingwonsong) [reply test/test_data/BUILD]: Done

File: test/wasm_test.cc (hunk 22 above)
Comment (PiotrSikora) [reply test/wasm_test.cc:17]: You need to keep external include (`gtest/gtest.h`) separately from local include (`test/utility.h`), i.e.
```
#include "include/proxy-wasm/wasm.h"

#include <unordered_set>

#include "gtest/gtest.h"

#include "test/utility.h"
```

File: test/utility.cc [hunk 23]
```diff
@@ -16,6 +16,8 @@
 
 namespace proxy_wasm {
 
+std::string TestContext::global_log_;
```
Comment (PiotrSikora) [inline test/utility.cc:19]: Does this need to be declared both here and in `TestContext`?

File: test/wasm_test.cc (hunk 22 above)
Comment (ingwonsong) [reply test/wasm_test.cc:17]: Oh. my mistake. Fixed.

**Review by PiotrSikora — COMMENTED** [review]
All three `YamlLoadFromFileWasmInvalidConfig` tests in Envoy are failing because of this change, which is a bit unexpected. Could you take a look?

Ideally, you would open a draft PR in Envoy pointing to this unmerged branch.

File: test/utility.cc (hunk 23 above)
Comment (ingwonsong) [reply test/utility.cc:19]: Because it is `static`, it needs to be declared in `utility.h` and defined in `utility.cc`.

**Comment by ingwonsong** [comment]
> All three `YamlLoadFromFileWasmInvalidConfig` tests in Envoy are failing because of this change, which is a bit unexpected. Could you take a look?
> 
> Ideally, you would open a draft PR in Envoy pointing to this unmerged branch.

Sure. will take a look

File: .github/workflows/test.yml [hunk 24]
```diff
@@ -56,6 +56,18 @@ jobs:
         --config=clang
         -c opt
         $(bazel query 'kind(was.*_rust_binary, //test/test_data/...)')
+        $(bazel query 'kind(_optimized_wasm_cc_binary, //test/test_data/...)')
+
```
Comment (PiotrSikora) [inline .github/workflows/test.yml:60]: Nit: trailing whitespaces.

File: .github/workflows/test.yml [hunk 25]
```diff
@@ -56,6 +56,18 @@ jobs:
         --config=clang
         -c opt
         $(bazel query 'kind(was.*_rust_binary, //test/test_data/...)')
+        $(bazel query 'kind(_optimized_wasm_cc_binary, //test/test_data/...)')
+    
+    # Currently, in Github Action, "Bazel build step" creates `wasm_canary_check.wasm` directory as a temporal directory.
+    # Since the name of this directory has "*.wasm" pattern, the step "Mangle build rules to use existing test data" fails.
+    # So, here, we clear out the directories in test_data.
```
Comment (PiotrSikora) [inline .github/workflows/test.yml:63]: Nit: trailing whitespace.

**Comment by ingwonsong** [comment]
@PiotrSikora 

Can you check out the behavioral difference in the fixed envoy test cases?
https://github.com/envoyproxy/envoy/pull/22469/files 

TL;DR - now we always apply canary, so the creation of FilterConfig or AccessLog should raise the exception always when the invalid config is configured.

**Review by PiotrSikora — APPROVED** [review]
Thanks!
